### PR TITLE
Refactor blocks implementation

### PR DESCRIPTION
**Pull request by alyosha**
# Summary
The current use of `Block`/`BlockElement` interfaces as initially introduced in #480 is proving problematic as Go interfaces have no underlying data structures into which we can unmarshal their corresponding JSON values. At the moment, attempts to unmarshal `Block`/`BlockElement` result in an error message similar to the one seen in issue #493. 

In addition to the above issue, Slack is now returning dynamic JSON values for each of the following keys, so some sort of type assertion and custom unmarshalling seems necessary regardless.
- `actions` (i.e. attachment or block)
- `blocks`
- `block_elements`

(not an exhaustive list)

The interfaces themselves still have a valid use case in simplifying the process of creating/sending Block messages, so I have left them more or less as is (still exported).

I have added new structs for use in unmarshalling/accessing the JSON response and implemented custom unmarshalling/marshalling logic for each where necessary. 

I wanted to get a working PR up ASAP and I realize this is still rather rough. I will be adding more tests and cleaning up in the next few days, but in the meantime please feel free to leave feedback. Also, if there is an easier way to accomplish the same thing, I am all ears.

### CODE REVIEW

File: block_conv.go
Comment (james-lawrence) [inline block_conv.go:39]: this loses ordering of the blocks doesn't it?

the blocks type seems unncessary we want this function but have it return an array of the Block interface.
```golang
func foo(data []byte) (blocks []block, err error) {
        type union struct {
           Type string `json:"type"`
        }
	err := json.Unmarshal(data, &raw)
	if err != nil {
		return err
	}
        for _, r := range raw {
            u := union{}
            json.Unmarshal(r, &u)

            switch(u.Type) {
            case "actions":
                var tmp ActionBlock
                json.Unmarshal(r, &tmp)
                blocks = append(blocks, tmp)
            }
        }
        return blocks, nil
}
```

File: block_conv.go
Comment (james-lawrence) [inline block_conv.go]: assertions are unnecessary, though I believe this method goes away anyways
```golang
 switch x := block.(type) {
 case *ActionBlock: b.ActionBlocks = append(b.ActionBlocks, x);
```

File: block_conv.go
Comment (alyosha) [reply block_conv.go:39]: @james-lawrence if we return an array of the block interface, how do you actually access the fields in that block? I want to be able to unmarshal the callback response, and then access the data in it for subsequent tasks, hence the `Blocks` type

**Comment by khayyamsaleem** [comment]
can some examples of handling the block interaction callbacks be added to the examples library?

**Comment by alyosha** [comment]
@khayyamsaleem sure I will add some examples later, but the basic idea is that any blocks present will be unmarshalled into their type's corresponding slice in the `Blocks` struct 👍

File: block_conv.go
Comment (james-lawrence) [reply block_conv.go:39]: type switch. we're effectively talking about sum types here. which generally use type switches to determine the current type of an element.

I think its more important to maintain order of the messages than it is for each to be explicitly typed in the results.

**Comment by khayyamsaleem** [comment]
@alyosha is it possible to receive the Block Interactions as RTM events?

File: block_conv.go
Comment (alyosha) [reply block_conv.go:39]: @james-lawrence that's a fair point about ordering that I hadn't really considered. I suppose if you wanted to replace a message with the same block order + some small modifications, it would be pretty inconvenient to not know the original order of the blocks.

just to confirm, you are suggesting that library users implement the type switch themselves directly, yes?

**Comment by alyosha** [comment]
@khayyamsaleem not sure I fully understand your question, but blocks are interactive components so the callbacks from interactions with them should all be forwarded to the callback URL you specify in your app dashboard

File: block_conv.go
Comment (alyosha) [reply block_conv.go:39]: will update the PR after confirming this point and ask for re-review 👍

Comment (james-lawrence) [reply block_conv.go:39]: Yes, I'm suggesting that library users implement the type switch themselves as they know exactly how they want to treat the blocks.

**Comment by james-lawrence** [comment]
@alyosha as an aside I did a lot of internal refactoring last week of the internal api's in the library. I doubt much of it impacts you but you may want to rebase off master.

File: block_conv.go
Comment (alyosha) [reply block_conv.go:39]: All right I will adapt this and leave some examples on how to unmarshal, thanks for the response

**Comment by alyosha** [comment]
sorry for the wait all, confirmed how to move forward with the dynamic unmarshalling so I will try and get this fixed/ready for re-review by @james-lawrence this week
🙏

File: examples/blocks/blocks.go
Comment (alyosha) [inline examples/blocks/blocks.go]: oops 👀 🤔

**Comment by alyosha** [comment]
@james-lawrence not sure if something is up with the CI, but keeps failing with the following message:

```
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.

Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received

The build has been terminated.
```

Can run tests locally without issue if I switch all the package names to point to my fork and not the upstream

**Comment by james-lawrence** [comment]
interesting, will take a look this weekend.

**Comment by alyosha** [comment]
closing and reopening PR to try and address CI issue
https://github.com/DefinitelyTyped/DefinitelyTyped/issues/27185#issuecomment-404503572

**Comment by alyosha** [comment]
I am now able to reproduce the hanging test behavior locally and am continuing to debug.

**Comment by alyosha** [comment]
CI issue resolved just making a few more convenience changes and adding examples, will re-request review tomorrow

File: block_element.go
Comment (james-lawrence) [inline block_element.go]: lets keep acronyms in caps.

File: block_context.go
Comment (james-lawrence) [inline block_context.go]: s/ContextElementSet/Elements/

File: block_context.go
Comment (james-lawrence) [inline block_context.go]: s/[]MixedElement/...MixedElement/

File: block_element.go
Comment (alyosha) [reply block_element.go]: will update this for all occurrences but probably tomorrow since I'm about to sleep

File: block_context.go
Comment (james-lawrence) [reply block_context.go]: actually does this method need to exist? or can we just use `...MixedElement` for NewContextBlock

File: block_context.go
Comment (alyosha) [reply block_context.go]: agree `Elements` is better here. I went with `BlockSet` in the case of `Blocks` because I couldn't really come up with anything better, but totally open to suggestions

File: block_conv.go
Comment (james-lawrence) [inline block_conv.go]: define a type inside this method for this unmarshalling. generally avoid maps were possible.
```
type sumtype struct {
  blockType string `json:"type"`
}
s := sumtype{}
json.Unmarshal(r, &s)
```

File: block_object.go
Comment (james-lawrence) [inline block_object.go]: this function seems entirely unnecessary, also seems unused now.

File: block_context.go
Comment (alyosha) [reply block_context.go]: will look into this tomorrow ^ :eyes:

File: block_conv.go
Comment (james-lawrence) [reply block_conv.go]: you could also define this type and embedded it in the block types. ./shrug depends how much it'll be used.

File: block_object.go
Comment (alyosha) [reply block_object.go]: might indeed be unnecessary now, will confirm tomorrow 👍

Comment (alyosha) [reply block_object.go]: removed

File: block_conv.go
Comment (alyosha) [reply block_conv.go]: changed from map to struct

File: block_context.go
Comment (alyosha) [reply block_context.go]: indeed did not need to exist; removed and updated as suggested

File: block_context.go
Comment (alyosha) [reply block_context.go]: Maybe `LayoutBlocks` would be a better naming choice than `BlockSet`?

File: block_element.go
Comment (alyosha) [reply block_element.go]: done ^

**Comment by alyosha** [comment]
@james-lawrence updated based on your feedback and added a type switch example to `examples/block.go`. left the conversations unresolved in case that was easier but feel free to resolve them if you're good 👍

**Comment by harrisonturton** [comment]
This is a big blocker for me - glad to see it being resolved!

**Comment by alyosha** [comment]
@harrisonturton yeah this is a blocker for all of my personal projects as well at the moment (no pun intended). sorry for the delay on getting this done 🙏

**Comment by alyosha** [comment]
@james-lawrence know how busy you are but please give this another look if you have a second 🙏

**Comment by alyosha** [comment]
@james-lawrence friendly ping 🙃

**Comment by james-lawrence** [comment]
sorry, been swamped w/ non-software related issues. I'll get to this tonight.

**Comment by dolfelt** [comment]
I noticed when using this branch that on marshaling back to JSON there is an unnecessary blocks key that causes errors when responding to interaction requests.

```json
{
    "replace_original": false,
    "delete_original": true,
    "blocks": {
        "blocks": null
    }
}
```

When adding `omitempty` to the `BlockSet` it fixes the issue.

```go
type Blocks struct {
	BlockSet []Block `json:"blocks,omitempty"`
}
```

File: block.go
Comment (james-lawrence) [inline block.go:28]: omit empty to prevent valdiation errors as noted in the other comment.

**Review by james-lawrence — APPROVED** [review]
Okay, so I'm going to merge this as is. even w/ the follow up work required just to unblock it.

and we'll go from there. plenty of stuff I can see to improve but should block hte bulk of the work.

please follow up with a PR fixing the validation erorr.

**Comment by alyosha** [comment]
will follow up with the `omitempty` change tonight–thanks as always @james-lawrence and thanks also to @dolfelt

**Pull request merged**
